### DISCUSSION TITLE
How can I use Spatie Translatable?

### DISCUSSION THREAD

**Discussion by Eyvazov** [Z3d0X/filament-fabricator#186] [Q&A] [answered]
Hello. I am using Filament Fabricator in my Laravel project. I added the Filament Spatie Translatable plugin for the multi-language system. I could use it everywhere I wanted except for pages, but I could not use it in Fabricator pages. Can you help me on how to use it?

**Comment by Voltra** [comment, marked as answer]
In my personal project, I do use the `filament/spatie-laravel-translatable-plugin` package (haven't fully tried the official community fork yet), and here's it how I made it work:

- I have a few patches to make it work properly (replaces `getState()` to `getState(false)`, if you go on the package's github you can find my PRs and extract them as patches)
- I have a `LocaleDetector` middleware the interecepts all requests and `App::setLocale($locale)` based on the URI
- I have a `App\Models\Overrides\Page` model that I set up in the package as the replacement (it extends from the package's `page` model)
  - I customize the `getDefaultUrlCacheArgs`, `getUrlCacheKey`,`getUrl` and `getAllUrlCacheKeysArgs` methods to expect a `locale` key to differentiate the pages
  - It has the `title`, `slug` and `blocks` columns in the `$translatable` field
- I have a `App\Filament\Resources\PageResource` that extends from the package's (same goes with all the pages) so I can setup the traits and other stuff
- I've added a migration to make `title` and `slug` properly translatable by turning them into `json` columns (like `blocks` already is)
- (optional) You could also disable the default routing in the config file and make your own (inspired by ours?) especially if you want full control on the arguments passed to the method that will resolve the page/page's route
